### ISSUE TITLE
Home page - change hero image upon refresh

### ISSUE THREAD

**Issue by joelworsham** [JacksonCollege/Primary-Theme#32]
_From @banksashleym on December 10, 2015 14:23_

Hi Joel,
Would it be possible on the home page to have the functionality to add additional spot for the hero image, so that upon refresh there would be a new hero image/story?

Thanks!

_Copied from original issue: JacksonCollege/Staff#16_

**Comment by joelworsham** [comment]
_From @brashrebel on December 10, 2015 21:4_

@banksashleym that's definitely possible. There's a number of ways to handle it though and reasons why it is worth considering them. Which image shown from the defined options could be:
- Totally random each page load
- Totally random for each unique visitor
- Different every X number of minutes, hours or days
- Different every time you change it (we can make this super easy if it will be needed often)
- Different based on some other criteria

An important question to ask yourself is: Why do you need to show alternate images? Does this help the user in some way? Or is it just for the sake of variety? I have no opposition to mixing it up, but I do assert that overdoing this can have negative consequences.

The most important concern when anything "random" comes up, is performance. We want the site to be fast and to do so we need to leverage multiple layers of caching and minimize complex database queries. Any query with `order => 'rand'` is going to be slow. In fact, WP Engine doesn't even allow that kind of query by default so we'd have to turn it on. A highly trafficked site with queries like that can actually take down a web server if you're not careful. But more than that, if we need a random image to show, then we can't cache the page or at least that part of it and that will seriously slow things down. So we need to ask ourselves is it better that a user load the home page in 0.5 seconds and see the same image as the previous visitor or should they load the home page in 3.5 seconds and see a different image? The reality is we can get the home page loading so much faster if we can cache the page which means the same image will be served.

That isn't to say we can't mix it up. We can invalidate the cache after a certain time period so a new image is shown after X amount of time which means no random queries and only a fraction of visitors will be served un-cached content. Or we can statically define the images for certain times. You could even set a unique image for each day of the week if you want. But every page load would be as inefficient as can possibly be done.

Additionally, it is worth considering what level of control you'd sacrifice by using a strictly random system. If a point came where you wanted to ensure visitors were all shown a specific image for some reason (seasonal or holiday related comes to mind), more work would be involved. Just another thought.

In summary, I believe that a truly random, every page load switch, would provide only a little value (virtually none to end users, only a little to JC marketing) and would come at a high performance cost. Because of this, I strongly recommend either a static image which you replace as often as you like or a series of images which swap at predefined times.

**Comment by joelworsham** [comment]
_From @banksashleym on December 11, 2015 12:57_

Thanks for the input! I think we wanted a little more variety in the hero image, so say that a user visited the site a few times a day over the course of a few days, it would be nice for them to see a little more variety. I trust your judgement and agree that maybe random isn't the best way to go?

I like the idea of setting the hero image to switch out at predefined times? Let me know your thoughts on that. I agree that it would be nice to have some sort of control over when the image was viewable, just in case we do need to have one specific message out there longer or alone.

Thanks!!

**Comment by joelworsham** [comment]
_From @brashrebel on December 11, 2015 14:45_

I think a mechanism for specifying when each of the pre-defined images are used would be a nice feature. You'd have lots of control and we'd minimize our performance impact. My professional opinion is that more variety in the hero image is something that will add very little value to the project but if we approach this the right way it won't introduce any negative consequences.

**Comment by joelworsham** [comment]
Do we have anything decided on this? @brashrebel @banksashleym

**Comment by banksashleym** [comment]
I think the last that Kyle and I spoke we figured we could add one or two more featured images and set them to swap out at pre-defined times? That way we could still have some variety among the hero images yet have control over when they are showing up?

**Comment by joelworsham** [comment]
:+1:

**Comment by joelworsham** [comment]
All set.

Me and Kyle discussed and decided the best way *for now* is to have the image be randomly selected every 2 days from a pool of images that you choose. Every 2 days a new image is selected randomly.

Note: I've added a button below where you add the images called "Flush Saved Hero Image". This takes you to the homepage, but also refreshes the hero image to a different, yet random image.

**Comment by banksashleym** [comment]
Ok sounds good...I talked to Dotty and Marilynn about this in length yesterday and they would prefer to - instead of doing a general JC image - feature our programs/relevant news items in this hero image spot. For instance there might be three hero images that we have, one for the Science program (linking to a science web feature page that Marilynn will write), one for Nursing, (linking to a science web feature page that Marilynn will write), and one highlighting our MLK event (linking to the events or news item). 

They really want to see a random refresh, and don't they don't have an issue with each of the three randomly refreshing, since each of the programs really have equal weight as far as exposure. If that is not advised, could we have them pre-timed so that they change switch every minute or few minutes? For instance every minute or so the science one would be visible, then the next minute the nursing story, then the next minute the MLK story? That way if we do need to increase the timing on one hero image we could do that? Let me know your thoughts on this.

**Comment by joelworsham** [comment]
The final decision for now is (for my reference):

- Use a CPT for controlling the images
- Use an explicit "featured" setting on each post for controlling which show on the home page
- Image / Link changes once per day, in the middle of the day
